arangod/Aql/ExecutionEngine.cpp

```diff
@@ -272,10 +272,35 @@ ExecutionEngine::~ExecutionEngine() {
   // tasks running on any dependent which could start another on the
   // current block.
   // The blocks are pushed in a reversed topological order.
-  for (auto it = _blocks.rbegin(); it != _blocks.rend(); ++it) {
-    (*it)->stopAsyncTasks();
+  {
+    // Note: We can use raw pointers here because we are not taking
+    // any responsibilty of the pointer. It is still managed by the _blocks
+    // vector. Also we are in the destructor here, so we have guaranteed,
+    // that no one else is cleaning up the blocks.
+    std::unordered_set<ExecutionBlock*> seenBlocks;
+    for (auto it = _blocks.rbegin(); it != _blocks.rend(); ++it) {
+      auto block = it->get();
+      if (ExecutionBlock* seenDependency =
+              block->isDependencyInList(seenBlocks);
+          seenDependency != nullptr && block->getPlanNode()->getType() != ExecutionNode::GATHER) {
+        // We have a dependency that has already been seen, we need to log this
+        // situation in theory this could lead to deadlocks. Some Blocks are
+        // fine we just want to see those here.
+        // Gather Nodes are known to violate this, but they are safe.
```

Comment:
Should we not do the warning for gather nodes then, to not spam the log with warnings?

Reply:
I would like to point you to the following line:

`seenDependency != nullptr && block->getPlanNode()->getType() != ExecutionNode::GATHER) {` 

;)